### PR TITLE
Merged in the S3-project branch.

### PR DESCRIPTION
**Pull request by pleblanc1976**
We'll have to think about how to make the CS and the SM repos coexist.  Right now SM is included as a submodule in the CS repo, but really all it needs is a header file from that repo to know how to construct & parse the messages between the two sides.  Including a whole repo is a heavy way to get that one file.  TBD.

Squashed commit of the following:

commit 62c7ab648028d15c888ee88db60ed0d9e6b62424
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Aug 5 16:35:59 2019 -0500

    Fixed a merge error.

commit 295c638c18c0efcd3f05808bad2761ce8ba0f04a
Merge: 99f8a4fe d30745d3
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Aug 5 16:34:45 2019 -0500

    Merge branch 'develop' into S3-project

commit 99f8a4feb7b29f563493bc4bccd00fde4be80ee1
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Jul 30 16:47:46 2019 -0500

    Update storage-manager ref.

commit 303d182d5c9a3fdd8ba199f2bc4afb743d2fe9c4
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 29 14:44:18 2019 -0500

    Removed a couple unused SM config params.  Added retry loop
    to getting a connection.

commit 00158c86ecacec6977552d50d1401917185b7b03
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 29 14:41:23 2019 -0500

    Minor improvement to what determines whether a file is 'local'
    or on some 'other' kind of filesystem.

commit 8190e44d1fa8a90b72594dfb284356ebf1b6f3e0
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 29 14:26:13 2019 -0500

    Fixed an old bug in the HDFS code in the BRM.  Journal file supposed
    to be opened for appending.

commit fa1ad8faca94f03cee4f0ae2e5323e608abe633b
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 22 17:04:41 2019 -0500

    Made procmon touch the journal file using idb file ops.  Not sure
    why this is necessary for load_brm to work.

commit f6b464f49097fcc6db5da54b33875f18fe34971c
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 22 16:44:01 2019 -0500

    A couple more hacks in procmgr.

commit 094e5b554ac2b7658ee03109999b477c6d5a0808
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 22 16:25:31 2019 -0500

    Changed debugging outputs.

commit 45d4a7536be3f500e66c7beffce300a48ba115ba
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 22 14:47:04 2019 -0500

    Need to make it eat a trailing \n char.

commit 1f09c533d522d1945353771a67aecc0d9cb021e6
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 22 14:12:41 2019 -0500

    more debugging printouts

commit ee81f96f4a869eff1b88cd6c77b85cda5d38beef
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 22 13:31:02 2019 -0500

    More debugging printouts.

commit 9f938447728d15eb1d9a2c602d06befc2c2be750
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 22 10:39:17 2019 -0500

    Still debugging brm transfer stuff.  Added some debugging printouts.

commit e8dd6443d62a4c5bab2f7e70bcab45fc91a99b5f
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 22 09:58:10 2019 -0500

    Replaced some of the shell stuff it was doing with IDB FS fcns.

    Still debugging...

commit 9daa44efa143d3307d4d182671518a541f4f7f2c
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Jul 19 17:08:29 2019 -0500

    Made procmgr init the IDB* stuff.

commit 52bd0ee967d9fcfae5746135781c28fb4c717f21
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Jul 19 15:52:01 2019 -0500

    Some logging changes for debugging.

commit be214a6fda6e0fa82da3ff69460b790ebdbc48ff
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Jul 19 14:54:33 2019 -0500

    Fixed a bug re the previous commit.

commit 1a7e7330169094826f890322dae5f9eced20bd2e
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Jul 19 14:33:29 2019 -0500

    Hacked IDB-IO ops into the getDBRMData() stuff procmgr does.

    Replaced the fstream stuff it did.

commit d0a576483e2eb7ed56aceb10cb7d25992ae88dbb
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Jul 19 09:54:54 2019 -0500

    Tentative fix for procmon d/ling DBRM files into data1.  That screws things
    up on shared filesystems, and storagemanager, which has sort-of a shared
    filesystem view.

commit 3118acc3d4e2a66adbaffedb17240c7bb23455ce
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jul 18 14:56:14 2019 -0500

    Made stale sockets reconnect.

commit 7d4f4068886f1d77b4ff23dda47a6e765f190a66
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jul 18 11:14:12 2019 -0500

    Changed the retry behavior in SocketPool to get load_brm to work
    when called by proc???.  Also made proc??? start/stop SM around
    save_brm runs.  There seems not to be a saner way to do that atm.

    So, on stop/restart/shutdown, it will do
    1) stop everything
    2) restart SM
    3) save_brm
    4) stop SM again

commit 43f7a71bad539fecbbeb62d7aa2d17568c4a30b5
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Jul 17 13:07:08 2019 -0500

    Disable force-killing of StorageManager.  Data corruption is possible
    ATM.

commit 4ec3cdc8b95633fbb7df6def9a00cf4804048e26
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 15 15:15:38 2019 -0500

    Fixed a bug where procmon would have a PID of 0 for StorageManager,
    and would end up sending SIGTERM to everything.

commit 3f532eb0596e174f93af13b13f8a10048aa8a20e
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jul 15 09:48:25 2019 -0500

    Possible fix for OAM madness when SM is killed externally.

commit b21b41077ebb8571188a87f9cb126bd57aebdc95
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Jul 12 13:48:36 2019 -0500

    Updated storage-manager ref.

commit 4c5866bccae0905cd5966ec6983f91794e14ffd7
Author: Ben Thompson <ben.thompson@mariadb.com>
Date:   Mon Jul 1 12:28:11 2019 -0500

    changes for ssize_t messaging

commit b30ee07318e9a62992dab51af3b3052693d9e960
Merge: eead61f2 95c4f128
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jun 24 15:41:10 2019 -0500

    Merge remote-tracking branch 'upstream/develop' into S3-project

commit eead61f238fe76afc7e6e64ba7a9367847d27b9a
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jun 13 09:26:10 2019 -0500

    Fixed the broken build, flipped the switch to put BRM files in SM.

commit a79b3131b5e66eacc809819ffdf1e971d1f09206
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jun 13 08:18:21 2019 -0500

    A little modding to get oam to play nice with storagemanager wrt
    BRM files.

commit 42b67765981533496b32130fe0e8492eba2d96b1
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Jun 12 13:32:24 2019 -0500

    Changed how procmon kills StorageManager to allow it to shutdown
    gracefully.

commit 753139a933f758c0439e0f4e92989795f764f124
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Jun 12 13:07:22 2019 -0500

    Checked for references to DBRootStorageType, made a couple changes
    I suspect are necessary to make failover work.

commit 366cd406bad29e5b20d17fb89b6dba61e9a20206
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Jun 12 10:00:11 2019 -0500

    Cleaned up the diff vs S3-project.

commit 4ae09352d2613bfd41cfb2a0e5d5870a2c9583ce
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Jun 12 09:37:45 2019 -0500

    Got postConfigure to set up storagemanager correctly and disabled
    the disk-check stuff for SM clusters.

commit 0ae0761e1a9659ec52a6a2cbd36099d217e21095
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jun 10 13:11:04 2019 -0500

    Removed the Location config value for SM; redundant with ProcessConfig.xml

commit f34f785f8a5eaa6ec0f95e18c4b2a6dc27cc098b
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jun 10 13:02:56 2019 -0500

    Made OAM suppress its knowledge of StorageManager if it is disabled.

commit d454f84b82972210aa40db482c9995f0486701cb
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jun 10 09:41:03 2019 -0500

    Cleaned up the hacks made so far.

commit 6193d2cda2f2ee261c8366396fdf6c976fda876a
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jun 10 08:10:10 2019 -0500

    Made mcsadmin and postConfigure print a warning if SKIP_OAM_INIT is set.

commit 15c256b01195cbb433ffdc67f79e13ec548444e5
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jun 6 15:47:15 2019 -0500

    Checkpointing some stuff.  Doesn't work yet.

commit 98c9cd7b22109c8c139dc205e46ec00643866d0e
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jun 6 08:36:40 2019 -0500

    Checkpointing some experimental changes to get storagemanager running.

commit 42b79bd5e77a76e7dc9b73e9b4d24348e8711ebd
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jun 3 09:43:05 2019 -0500

    Added a couple config params for storagemanager.

commit 3d67df18bf3132ce32cb74683e932107e6de5dc7
Merge: 3c32f6d8 355e2f91
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri May 17 13:24:09 2019 -0500

    Merge branch 'develop' of github.com:mariadb-corporation/mariadb-columnstore-engine into S3-project

commit 3c32f6d888df7858ca49759458e884dee574cbc2
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Apr 25 14:15:25 2019 -0500

    Added SMDataFile::fallocate(), it's just a call to truncate() for now.

    Updated ref of storage-manager subproject.

commit 1dd328bbb1e4c61848d03091a67e5d9948b0ab99
Merge: 08923813 f99395cf
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Apr 25 13:40:22 2019 -0500

    Merge branch 'develop' into S3-project

commit 08923813cbea0207e11752c61ab13245c2b49730
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Apr 23 16:01:19 2019 -0500

    Fixed the error path in SMDataFile read() and write()

commit 46163c129ac9216ace94d15b272a3773e4f7685b
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Apr 10 17:53:05 2019 -0500

    Neutered all of the excessive snapshotting that was going on.

commit dfc29fa833613c2e4ccbe34033e6c0c2226c0282
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Apr 10 15:01:29 2019 -0500

    Eventually found the problem with rename().  Was calling
    ::unlink() instead of our unlink/remove.

commit c0f93d904a80e75c2020c7fb685597b286e58f6e
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Apr 9 17:34:30 2019 -0500

    Fixed a bug where the BRM would open the journal for reading when
    it should be have open for writing.  Not sure how that ever worked.

commit 6e6cfdb59b0cab82abf180a3b66e8c0196ecc34b
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Apr 9 15:07:00 2019 -0500

    The BRM structs aren't using IDBDataFile consistently.

    Testing whether forcing them to use IDBDataFiles breaks anything.

commit 2b9e7facd28a2c22167aaa7c711c76a348b9e04b
Merge: e060cce4 11bec446
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Apr 9 10:28:31 2019 -0500

    Merge remote-tracking branch 'upstream/develop' into S3-project

commit e060cce496975cdd23aa68841bc9d1c3e83e2a76
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Apr 9 10:27:32 2019 -0500

    Updated ref for storage-manager.

commit 2a510c364118fc33146116e1bb266958263c169d
Merge: a9ce25c4 668f83c1
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Feb 11 18:20:45 2019 -0600

    Merge branch 'S3-project' of github.com:pleblanc1976/mariadb-columnstore-engine into S3-project

commit a9ce25c4f68adf60ecdeb709a5839d7cbcc43d82
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Feb 11 18:20:11 2019 -0600

    Phase 1 of cleaning up the messaging code.

commit 668f83c1635a29da5f12e865adde75dbe4f6fb32
Author: Ben Thompson <ben.thompson@mariadb.com>
Date:   Fri Feb 8 15:15:37 2019 -0600

    end_to_end_test initial commit.

commit d1297f31a207e7d6a62b45551b1f1dca51598d21
Merge: 5528aba6 46cc3441
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Feb 8 14:08:06 2019 -0600

    Merge branch 'develop-1.2' into S3-project

commit 5528aba68a278afd7d56b413f47d5fd2ca127da0
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Feb 7 10:01:05 2019 -0600

    Updated the ref for storage-manager.  Test000 passes here.

commit 4f26ce089cb5933e110dd302aaeea72513894448
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Feb 7 09:57:36 2019 -0600

    Made socket pool write the header in one write call.  Added some debugging
    printouts (commented).

commit 92f609f718ae505eb7b2fd756968f181b28116f5
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Feb 6 09:49:32 2019 -0600

    Fixed a bug in IDBPolicy that prevented using the system
    w/o plugins.

    A couple other random things.

commit c9a9bf98237cdb59049aa5eb9fcce5e9fb8501dc
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Feb 5 14:20:17 2019 -0600

    Update subproject ref.

commit 8c1802dccde9a963b6cbf4f8080c0a5e124f7481
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Feb 4 10:57:31 2019 -0600

    Update the storage-manager ref.

commit b6d28b9a84c582d8109319a355bdaba9d4d04087
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Feb 4 10:56:21 2019 -0600

    Fixed a bug processing a directory listing.

commit 0d86651fae05c30aaa414e8dde0a57793237815a
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Feb 4 10:44:17 2019 -0600

    Added a missing file.

commit d97a570b866a5ad3c0d6a74d8178318e030101ec
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Feb 4 10:41:21 2019 -0600

    Made the cloud IO lib a plugin to the IDB filesystem stuff.
    It loads correctly.

commit 8ad318816243e47fb355f93d5c717adba10f95d9
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Feb 1 13:13:30 2019 -0600

    Updated the storage-manager ref.  Made SMComm send absolute
    paths rather than relative ones.

commit a9a5c2ebb70d55655007793ee87a2816712059ca
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Jan 29 09:53:18 2019 -0600

    Reordered a couple params.

commit 918e10a9ffdcfd982624e6b6c71c9dec188b702d
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jan 28 17:27:06 2019 -0600

    Modified the sequence some params were serialized to allow
    fewer network/socket reads.

commit 4583b2e8dc5d00dfbdb585c2f7e00dc7a45acc70
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jan 24 17:34:20 2019 -0600

    Added SMDataFile methods to the list tested.

commit a2c789d52a4176c09cf4755116624b1789eb0104
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jan 24 17:05:09 2019 -0600

    Added remaining methods in SMFileSystem to the test.

commit acb464618c38dba689e6f86854da62836f1d10c8
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Thu Jan 24 16:52:26 2019 -0600

    Wrote a component test, which has a stand-in server thread
    in place of an actual StorageManager.  The server thread
    returns the same error over and over again.

    Used that test to debug some basic comm stuff & find
    errors in error handling.

commit f063f78242d3dc429fa6b6177aa7456bbf33f3af
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Jan 23 14:12:57 2019 -0600

    Fixed some linker errors, integrated the cloudio factory
    into idbfactory.

commit 099c724ced70e011772d4418179373d37307e37f
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Jan 23 13:05:07 2019 -0600

    Update storage-manager ref.

commit 51bb9f3050c10f8bc53a617863d3c853643ad055
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Wed Jan 23 13:02:53 2019 -0600

    Added licensing preamble.  Got it to build.

commit 88273bfc1455dcdd9b48614d790e810506349299
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Jan 22 16:12:49 2019 -0600

    Finished the first cut.  I assume it doesn't build yet.

commit 3d58581945a1a7fdaaaac3aa030b086394ad7b53
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Jan 22 16:04:32 2019 -0600

    Update storage-manager ref.

commit 2e5d6db0c73748f11fd7d5d524ef564e4ac8f7bb
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Jan 22 10:17:18 2019 -0600

    Forgot another small thing.  I'm a dummy.

commit 259060fb29fd3fda5ff9b54c87c72df6cf725933
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Jan 22 10:11:50 2019 -0600

    Forgot something in the bytestreampool code.

commit d92be01ee2194c34b2f3e9a5712a77d6fe219a65
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Tue Jan 22 10:09:08 2019 -0600

    Added a ByteStreamPool class for cases where we'd otherwise
    have to do something complicated, or constantly create and dispose
    of ByteStreams.

commit d84dcb9ccc52b64d294dbbd3b14b8ddf270ff9fe
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Mon Jan 21 16:41:04 2019 -0600

    Checkpointing, not ready to build yet.

commit d53471fc759ae3948cf9156b21ba8868b76b60fd
Author: Patrick LeBlanc <patrick.leblanc@mariadb.com>
Date:   Fri Jan 18 10:19:14 2019 -0600

    Checkpointing some stuff.  No way it'll build yet.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "utils/cloudio/storage-manager"]
+	path = utils/cloudio/storage-manager
+	url = git@github.com:pleblanc1976/storage-manager
```
Comment (LinuxJedi) [inline .gitmodules:3]: Can we please:

1. Move this to a more official repo (I can get one created if needed)
2. Add appropriate license headers to the source files in that repo (I can help with BSL headers if needed)

Comment (pleblanc1976) [reply .gitmodules:3]: Yes both of those need to be done.  Should we do that in this PR, or add a ticket + PR for follow-on mods?

**Comment by pleblanc1976** [comment]
Buildbot is failing, looks like it is outside of this PR.  does develop build right now?

**Comment by drrtuy** [comment]
I don't think so given that convergence MCOL-2178 had been merged.



чт, 8 авг. 2019 г., 3:01 Patrick LeBlanc <notifications@github.com>:

**Review by LinuxJedi — CHANGES_REQUESTED** [review]
Given conversations yesterday please put GPL headers on the storage-manager code and add it to this tree rather than a separate repo.

**Comment by pleblanc1976** [comment]
just migrated the SM repo into the CS one, will retract this one and make One Big PR with both parts.

**Pull request closed without merging**